test/AppNotificationTests/BaseTestSuite.cpp

```diff
@@ -0,0 +1,640 @@
+
+#include "pch.h"
+
+#include "AppNotification-Test-Constants.h"
+#include "AppNotificationHelpers.h"
+#include "BaseTestSuite.h"
+
+using namespace WEX::Common;
+using namespace WEX::Logging;
+using namespace WEX::TestExecution;
+
+using namespace winrt::Windows::ApplicationModel::Activation;
+using namespace winrt::Windows::ApplicationModel::Background;
+using namespace winrt::Windows::Foundation;
+using namespace winrt::Windows::Foundation::Collections;
+using namespace winrt::Windows::Management::Deployment;
+using namespace winrt::Windows::Storage;
+using namespace winrt::Windows::System;
+using namespace winrt::Microsoft::Windows::AppNotifications;
+using namespace AppNotificationHelpers;
+
+void BaseTestSuite::ClassSetup()
+{
+    ::Test::Bootstrap::SetupPackages();
+}
+
+void BaseTestSuite::ClassCleanup()
+{
+    ::Test::Bootstrap::CleanupPackages();
+}
+
+void BaseTestSuite::MethodSetup()
+{
+    ::Test::Bootstrap::SetupBootstrap();
+
+    bool isSelfContained{};
+    VERIFY_SUCCEEDED(TestData::TryGetValue(L"SelfContained", isSelfContained));
+
+    if (!isSelfContained)
+    {
+        ::WindowsAppRuntime::VersionInfo::TestInitialize(::Test::Bootstrap::TP::WindowsAppRuntimeFramework::c_PackageFamilyName);
+        VERIFY_IS_FALSE(::WindowsAppRuntime::SelfContained::IsSelfContained());
+    }
+    else
+    {
+        ::WindowsAppRuntime::VersionInfo::TestInitialize(L"I_don't_exist_package!");
+        VERIFY_IS_TRUE(::WindowsAppRuntime::SelfContained::IsSelfContained());
+    }
+    EnsureNoActiveToasts();
+}
+
+void BaseTestSuite::MethodCleanup()
+{
+    if (!m_unregisteredFully)
+    {
+        UnregisterAllWithAppNotificationManager();
+    }
+
+    ::WindowsAppRuntime::VersionInfo::TestShutdown();
+    ::Test::Bootstrap::CleanupBootstrap();
+}
+
+void BaseTestSuite::RegisterWithAppNotificationManager()
+{
+    AppNotificationManager::Default().Register();
+    m_unregisteredFully = false;
+}
+
+void BaseTestSuite::UnregisterAllWithAppNotificationManager()
+{
+    AppNotificationManager::Default().UnregisterAll();
+    m_unregisteredFully = true;
+}
+
+void BaseTestSuite::VerifyRegisterActivatorandUnregisterActivator()
+{
+    RegisterWithAppNotificationManager();
+    AppNotificationManager::Default().Unregister();
+}
+
+void BaseTestSuite::VerifyFailedMultipleRegister()
+{
+    RegisterWithAppNotificationManager();
+    VERIFY_THROWS_HR(AppNotificationManager::Default().Register(), E_INVALIDARG);
+}
+
+void BaseTestSuite::VerifyUnregisterAll()
+{
+    RegisterWithAppNotificationManager();
+    UnregisterAllWithAppNotificationManager();
+}
+
+void BaseTestSuite::VerifyUnregisterTwice()
+{
+    RegisterWithAppNotificationManager();
+    AppNotificationManager::Default().Unregister();
+    VERIFY_THROWS_HR(AppNotificationManager::Default().Unregister(), E_NOT_SET);
+}
+void BaseTestSuite::VerifyToastSettingEnabled()
+{
+    VERIFY_ARE_EQUAL(AppNotificationManager::Default().Setting(), AppNotificationSetting::Enabled);
```

Comment:
Can we have a negative test?

Reply:
Do you know of a way to programmatically set the AppNotificationSetting? I would love to add a negative case :)

Reply:
If you follow the bread crumbs to product code, I believe there is a way to set group policy via registry.

Reply:
I looked at the code. We definitely need a negative test but setting it is not trivial. We may need a test hook of some sort. LEts opene up a seperate work item for a negative test.

Reply:
https://github.com/microsoft/WindowsAppSDK/issues/2587